File: biddo_web/src/core/domain/bid.ts
import { makeAutoObservable } from 'mobx'
import { Account } from './account'
import { LatLng } from './auction'

export class Bid {
  id: string
  bidder?: Account
  location?: LatLng
  locationPretty?: string
  description?: string
  isRejected?: boolean
  rejectionReason?: string
  isAccepted?: boolean
  initialCurrencyId?: string
  price?: number

  createdAt?: Date
  updatedAt?: Date

  constructor(params: Record<string, unknown> = {}) {
    this.id = params.id as string
    this.bidder = params.bidder as Account
    this.location = params.location as LatLng
    this.locationPretty = params.locationPretty as string
    this.description = params.description as string
    this.isRejected = params.isRejected as boolean
    this.rejectionReason = params.rejectionReason as string
    this.isAccepted = params.isAccepted as boolean
    this.price = params.price as number
    this.initialCurrencyId = params.initialCurrencyId as string
    this.createdAt = params.createdAt as Date
    this.updatedAt = params.updatedAt as Date

    makeAutoObservable(this)
  }

  static fromJSON(data: Record<string, unknown>) {
    const locationLat = data.locationLat as number
    const locationLong = data.locationLong as number
    const locationLatLng =
      locationLat != null && locationLong != null
        ? new LatLng({ lat: locationLat, lng: locationLong })
        : null

    const bidder = data.bidder ? Account.fromJSON(data.bidder as Record<string, unknown>) : null

    return new Bid({
      id: data.id,
      bidder,
      location: locationLatLng,
      locationPretty: data.locationPretty ?? '',
      description: data.description,
      isRejected: data.isRejected,
      rejectionReason: data.rejectionReason,
      isAccepted: data.isAccepted,
      price: data.price ? parseFloat(data.price.toString()) : 0,
      initialCurrencyId: data.initialCurrencyId,
      createdAt: data.createdAt ? new Date(data.createdAt.toString()) : null,
      updatedAt: data.updatedAt ? new Date(data.updatedAt.toString()) : null,
    })
  }
}

File: biddo_client/lib/core/models/bid.dart
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'account.dart';

class Bid {
  String? id;
  Account? bidder;
  LatLng? location;
  String? locationPretty;
  String? description;
  bool? isRejected;
  String? rejectionReason;
  bool? isAccepted;
  double? price;
  DateTime? biddedAt;
  String? initialCurrencyId;

  Bid({
    this.id,
    this.bidder,
    this.location,
    this.description,
    this.isRejected,
    this.locationPretty,
    this.rejectionReason,
    this.isAccepted,
    this.biddedAt,
    this.price,
    this.initialCurrencyId,
  });

  static Bid fromJSON(dynamic data) {
    var bidder =
        data['bidder'] != null ? Account.fromJSON(data['bidder']) : null;

    var locationLat = data['locationLat'] != null
        ? double.parse(data['locationLat'].toString())
        : null;
    var locationLong = data['locationLong'] != null
        ? double.parse(data['locationLong'].toString())
        : null;
    var locationLatLng = locationLat != null && locationLong != null
        ? LatLng(locationLat, locationLong)
        : null;

    return Bid(
      id: data['id'],
      bidder: bidder,
      description: data['description'],
      price:
          data['price'] != null ? double.parse(data['price'].toString()) : 0.0,
      isRejected: data['isRejected'],
      locationPretty: data['locationPretty'] ?? '',
      rejectionReason: data['rejectionReason'],
      isAccepted: data['isAccepted'],
      biddedAt: DateTime.parse(data['createdAt']),
      location: locationLatLng,
      initialCurrencyId: data['initialCurrencyId'],
    );
  }
}
